File: day6/code.R
input <- readLines("day6/input.txt")

# part 1
reset_seq <- function(seq){
    if (any(seq == -1)){
        new_add = rep(8,sum(seq == -1))
        seq[which(seq == -1)]=6
        seq = c(seq, new_add)
    }
    seq
}

days <- 80
initial_timer <- as.numeric(strsplit(input[1], ",")[[1]])
seq <- initial_timer
for (day in 1:days){
    seq <- reset_seq(seq - 1) 
}

length(seq) # 352872

# part 2
initial_timer <- as.numeric(strsplit(input[1], ",")[[1]])
counts <- c(0,tabulate(initial_timer, nbins=8))

days=256
for (i in 1:days) {
    n0 <- counts[1]
    tmp <- counts[-1]
    tmp[7] = tmp[7] + n0
    tmp[9] = n0
    counts <- tmp
}
format(sum(counts), scientific=F) # 1604361182149

File: day6/code.py
with open("day6/input.txt", "r") as fh:
    lines = fh.readlines()

initial_timer = lines[0].rstrip().replace(" ","").split(",")
initial_timer = [int(i) for i in initial_timer]

# part 1
def reset_seq(seq):
    reset_bl = [i==-1 for i in seq]
    if (any(reset_bl)):
        reset_n = sum(reset_bl)
        for i,x in enumerate(reset_bl):
            if x:
                seq[i] = 6
        seq = seq + [8] * reset_n
    return seq

days = 80
seq = initial_timer
for i in range(0,days):
    seq = reset_seq([x-1 for x in seq])

len(seq) # 352872


# part 2
counts=[]
for i in range(0,9):
    bl = [d==i for d in initial_timer]
    counts.append(sum(bl))

counts

for i in range(0,256):
    n0 = counts[0]
    tmp = counts[1:9]
    tmp[6] = tmp[6] + n0
    tmp.append(n0)
    counts = tmp
    
sum(counts) # 1604361182149
